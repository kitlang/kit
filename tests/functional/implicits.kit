abstract SpecialInt: Int;
abstract SpecialFloat: Float;

function f(a: SpecialInt, b: SpecialFloat, c: CString) {
    printf("hello %i %.1f %s!\n", a, b, c);
}

function main() {
    // numeric types aren't a great example of implicits; for the test to pass,
    // using abstracts so they don't unify with each other
    var a: SpecialInt = 1;
    var b: SpecialFloat = 2.0;
    var c: CString = "world2";

    using implicit a {
        f(3.5, "world");
    }

    // implicit order shouldn't matter
    using implicit b, implicit a {
        f("world");

        // nested implicit
        using implicit c f();
    }

    var d: Float = 2.5;
    var e: Float;
    using implicit d {
        e = 4 + (implicit Float);
    }
    printf("%.1f\n", e);

    // if we cast e to an Int and promote to SpecialInt, this works
    using implicit ((e as Int) as SpecialInt) {
        f(5.1, "world");
    }

    // this does not use the implicit
    using implicit (e as Int) {
        f(a, 5.2, "world");
    }
}
