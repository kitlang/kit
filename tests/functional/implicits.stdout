hello 1 3.5 world!
hello 1 2.0 world!
hello 1 2.0 world2!
6.5
hello 6 5.1 world!
hello 1 5.2 world!
